File: TypeScript/42.接雨水.ts
function trap(height: number[]): number {
    let ans: number = 0
    const n: number = height.length
    if (Math.floor(n) == 0)
        return ans
    const leftheight = new Array<number>(n)
    const rightheight = new Array<number>(n)
    leftheight[0] = height[0]
    rightheight[n-1] = height[n-1]
    for (let i = 1; i < n; ++i)
        leftheight[i] = Math.max(leftheight[i-1], height[i])
    for (let i = n-2; i >= 0; --i)
        rightheight[i] = Math.max(rightheight[i+1], height[i])
    for (let i = 0; i < n; ++i)
        ans += Math.min(leftheight[i], rightheight[i]) - height[i]
    return ans
};

File: Python/42.接雨水.py
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        ans: int = 0
        n: int = len(height)
        if height is None:
            return ans
        leftheight: List[int] = [0] * n
        rightheight: List[int] =[0] * n
        leftheight[0] = height[0]
        rightheight[n-1] = height[n-1]
        for i in range(1, n):
            leftheight[i] = max(leftheight[i-1], height[i])
        for i in range(n-2, -1, -1):
            rightheight[i] = max(rightheight[i+1], height[i])
        for i in range(0, n):
            ans += min(leftheight[i], rightheight[i]) - height[i]
        return ans
